### PR TITLE
Added BUYPersistence protocol to serialize a BUYCart to a property list.

### PR DESCRIPTION
**Pull request by jimspeth**
This resolves https://github.com/Shopify/mobile-buy-sdk-ios/issues/77 and works in my testing.  If there are any code-style changes, or other objects that I missed, please let me know.

### CODE REVIEW

**Comment by jimspeth** [comment]
Updated with a test.

File: Mobile Buy SDK/Mobile Buy SDK/Models/BUYCart.m [hunk 1]
```diff
@@ -115,4 +115,35 @@ - (NSDictionary *)jsonDictionaryForCheckout
 	return cart;
 }
 
+#pragma mark - BUYPersistence
+
++ (instancetype)instanceWithPlistDictionary:(NSDictionary *)dictionary
+{
+	if (![dictionary isKindOfClass:[NSDictionary class]]) {
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Models/BUYCart.m]: We would prefer not to use defensive programming techniques. Callers are responsible for providing valid data when invoking a method. If they provide invalid data, the program should crash.

File: Mobile Buy SDK/Mobile Buy SDK/Models/BUYCartLineItem.m [hunk 2]
```diff
@@ -61,4 +61,25 @@ - (NSUInteger)hash
 	return hash;
 }
 
+#pragma mark - BUYPersistence
+
++ (instancetype)instanceWithPlistDictionary:(NSDictionary *)dictionary
+{
+	BUYProductVariant *variant = [BUYProductVariant instanceWithPlistDictionary:dictionary[@"variant"]];
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Models/BUYCartLineItem.m]: You should probably not serialize the entire variant object in the cart. The variant details might change between persisting the cart and loading it. Store the variant's `identifier` instead.

File: Mobile Buy SDK/Mobile Buy SDK/Models/BUYImage.m [hunk 3]
```diff
@@ -43,4 +43,22 @@ - (void)updateWithDictionary:(NSDictionary *)dictionary
 	_updatedAtDate = [dateFormatter dateFromString:dictionary[@"updated_at"]];
 }
 
+#pragma mark - BUYPersistence
+
+- (NSDictionary *)plistDictionary
+{
+	NSMutableDictionary *dictionary = [[super plistDictionary] mutableCopy];
+	
+	if (self.src)           dictionary[@"src"] = self.src;
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Models/BUYImage.m]: All `if` (and `else`) statement bodies must be wrapped in curly braces, and on separate lines, for safe coding reasons.

File: Mobile Buy SDK/Mobile Buy SDK/Models/BUYObject.h [hunk 4]
```diff
@@ -33,7 +34,7 @@
  *
  *  You will generally not need to interact with this class directly.
  */
-@interface BUYObject : NSObject
+@interface BUYObject : NSObject <BUYPersistence>
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Models/BUYObject.h]: `BUYPersistence` is not an appropriate name for this protocol. `BUYPropertyListTransformable` would be better.

File: Mobile Buy SDK/Mobile Buy SDK/Models/BUYObject.mm [hunk 5]
```diff
@@ -262,4 +262,26 @@ - (NSUInteger)hash
 	return hash;
 }
 
+#pragma mark - BUYPersistence
+
++ (instancetype)instanceWithPlistDictionary:(NSDictionary *)dictionary
+{
+	if (![dictionary isKindOfClass:[NSDictionary class]]) {
```
Comment (bgulanowski) [inline Mobile Buy SDK/Mobile Buy SDK/Models/BUYObject.mm]: Defensive programming.

**Comment by bgulanowski** [comment]
Hi, @jimspeth. I left some comments on your code, but you will probably not want to jump into making any changes right now. We are working on a local data caching approach, and this is going in a different direction.

In the short term, I recommend you do the following:
- Do not adopt your plist protocol on `BUYObject`; adopt it specifically on the cart model classes. Many subclasses of `BUYObject` are not appropriate for storing in a file, i.e. the checkout object graph.
- Consider implementing protocol adoption in a category instead of the main class files. This will make merging with our incoming solution much less troublesome.

The fact is, however, that this solution will not be required. I recommend that you implement it in your application directly, instead of in the SDK.

**Comment by bgulanowski** [comment]
Thank you for your contribution, in any case! It's definitely a good idea.

**Comment by jimspeth** [comment]
Thanks for reviewing it!  Do you have a timeframe for when your local caching changes might be available?

**Comment by bgulanowski** [comment]
I hesitate to predict anything in case something changes. Less than a month is probably a safe bet, but I can't literally promise anything.

**Comment by jimspeth** [comment]
Understood, thanks for letting us know.

**Comment by nabeelahmed87** [comment]
So we are still waiting for that Method ?

**Comment by bgulanowski** [comment]
The new `BuyModelManager` protocol defines the interface for client-side caching, if you want to be consistent for all models.

Although, in this case, the `BUYCart` is not part of the storefront model, and never works with carts, so it may not buy you anything to use that interface.

On the other hands, carts now do support JSON encoding/decoding, so if you want, you can implement a very simple model manager that converts carts to and from JSON and saves them wherever you like.

If you have any further questions, please post them to the Shopify forums:
https://ecommerce.shopify.com/forums

**Comment by nabeelahmed87** [comment]
I am able unable to encode/decode cart. Can you please tell me the steps so that i can encode and decode the cart. Thanks for the help

**Comment by nabeelahmed87** [comment]
I am using this funcation for encoding buyCart.jsonDictionaryForCheckout() and got this in json 

({
    "line_items" =     (
                {
            price = 55;
            quantity = 2;
            "requires_shipping" = 1;
            title = "Fresh Without Shell (500g)";
            "variant_id" = 20325638465;
        },
                {
            price = 109;
            quantity = 3;
            "requires_shipping" = 1;
            title = "Fresh without Shell";
            "variant_id" = 18464205953;
        }
    );
})

and to decode if this is the proper way becasue some info is not encoded in it

**Comment by bgulanowski** [comment]
This ticket is closed. Github is for issue tracking (bugs and feature requests). Please move discussion to the forums: https://ecommerce.shopify.com/forums

**Pull request closed without merging**